### PR TITLE
Nintendo Switch

### PR DESCRIPTION
**Pull request by AJenbo**
This is an effort to upstream the Switch port created by @lantus (MVG) in a manner that maintains other ports and full credit via git commits. To do so some commits have been edited, mainly when the change is both temporary and destructive or could cause conflicts.

This will hopefully help to resolve https://github.com/lantus/devilution-nx/issues/7

MVG's video about the project:
https://www.youtube.com/watch?v=5tADL_fmsHQ&t=34s

This PR is currently in draft mode while we work out some outstanding issues regarding the merge:
- [x] Unify code where there doesn't need to be a difference between Switch and other ports
- [x] Make a separate `#ifdef` for controller code and switch code so it can be enabled independently
- [x] Create a readme with Switch relevant info in the switch folder
- [x] Cherry pick all commits from https://github.com/lantus/devilution-nx without breaking other platforms
- [x] Reduce changes in the Source/ folder to a minimum
- [x] See if manipulating `MouseX`/`MouseY` directly is needed if `SetCursorPos` is re-enabled
- [x] Remove libsodium dependency for Switch
- [x] Expose `DSBSTATUS_PLAYING` in a way that is picked up by the Switch build process
- [x] Fix macOS builds
- [x] Merge with the latest code from master
- [x] Fix escaping videoes via keyboard input
- [x] Fix SDL1 builds

New features compared to devilution-nx:
- All builds are now automatically compiled for the switch via CircleCI
- Map all keyboard keys
- Fix save game compatibility
- Add missing shadow in the credits
- Fix zooming in the minimap for some keyboards
- Fix typos
- Fix TCP code
- Correct directional sounds
- Fix multiplayer bug
- All code now matches the original Diablo (fixing a few in game bugs)
- Tons of code clean up and better code documentation
- Fix clang-format configuration
- Correct file encoding
- Fix crash in town with some compilers
- Correct crash when exiting menu
- Error messages implemented in GUI
- Much lower CPU usage
- FPS no longer locked to 20
- Many bug fixes in the controller implementation
- More responsive controller
- highlight both primary and secondary target

### CODE REVIEW

File: SourceX/miniwin/misc_msg.cpp [hunk 1]
```diff
@@ -221,10 +290,160 @@ WINBOOL PeekMessageA(LPMSG lpMsg, HWND hWnd, UINT wMsgFilterMin, UINT wMsgFilter
...
+				lpMsg->message = e.type == SDL_MOUSEBUTTONUP ? DVL_WM_LBUTTONUP : DVL_WM_LBUTTONDOWN;
+				lpMsg->lParam = (e.button.y << 16) | (e.button.x & 0xFFFF);
+				if (lpMsg->message == DVL_WM_LBUTTONUP)
+					lpMsg->wParam = keystate_for_mouse(0);
+				else
+					lpMsg->wParam = keystate_for_mouse(DVL_MK_LBUTTON);
+			}
+			break;
+		}
+		return true;
```
Comment (AJenbo) [inline SourceX/miniwin/misc_msg.cpp]: This appears to be what breaks skipping videos via keyboard.

File: SourceX/storm/storm.cpp [hunk 2]
```diff
@@ -485,6 +488,11 @@ BOOL SVidPlayBegin(char *filename, int a2, int a3, int a4, int a5, int flags, HA
 		deviceId = SDL_OpenAudioDevice(NULL, 0, &audioFormat, NULL, 0);
 		if (deviceId == 0) {
 			SDL_Log(SDL_GetError());
+#ifdef SWITCH
+			Mix_OpenAudio(22050, AUDIO_S16LSB, 2, 1024);
```
Comment (AJenbo) [inline SourceX/storm/storm.cpp]: Probably redundant if `SDL_Init` is reverted to also enable audio.

File: switch/mman.h [hunk 3]
```diff
@@ -0,0 +1,125 @@
+#ifndef MMAN_H
```
Comment (AJenbo) [inline switch/mman.h]: Move to 3rdParty folder

File: SourceX/miniwin/misc.cpp [hunk 4]
```diff
@@ -322,11 +335,21 @@ HWND CreateWindowExA(
...
 	}
 	atexit(SDL_Quit);
+#else
+	if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_JOYSTICK | SDL_INIT_GAMECONTROLLER) <= -1) {
+		SDL_Log(SDL_GetError());
+		return NULL;
+	}
+#endif
+
+	SDL_JoystickOpen(0);
```
Comment (AJenbo) [inline SourceX/miniwin/misc.cpp]: Missing error handeling

File: SourceX/miniwin/misc.cpp [hunk 5]
```diff
@@ -322,11 +335,21 @@ HWND CreateWindowExA(
...
+	SDL_GameControllerOpen(0);
```
Comment (AJenbo) [inline SourceX/miniwin/misc.cpp]: Missing error handeling

File: SourceX/miniwin/misc.cpp [hunk 6]
```diff
@@ -322,11 +335,21 @@ HWND CreateWindowExA(
...
     LPVOID lpParam)
 {
+#ifndef SWITCH
 	if (SDL_Init(SDL_INIT_EVERYTHING & ~SDL_INIT_HAPTIC) <= -1) {
 		SDL_Log(SDL_GetError());
 		return NULL;
 	}
 	atexit(SDL_Quit);
+#else
+	if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_JOYSTICK | SDL_INIT_GAMECONTROLLER) <= -1) {
```
Comment (AJenbo) [inline SourceX/miniwin/misc.cpp]:
```suggestion
 	if (SDL_Init(SDL_INIT_EVERYTHING & ~SDL_INIT_HAPTIC) <= -1) { 
```

**Comment by mewmew** [comment]
Now we just have to get a Switch console for our CI cluster :p

File: SourceX/storm/storm.cpp (hunk 2 above)
Comment (rsn8887) [reply SourceX/storm/storm.cpp]: On Switch at least, this is still necessary even after fixing SDL_Init. SDL_Mixer docs specifically forbid calling SDL_OpenAudioDevice while using mixer.

File: SourceX/miniwin/misc.cpp (hunk 6 above)
Comment (rsn8887) [reply SourceX/miniwin/misc.cpp]: Fixed in PR #187

File: SourceX/miniwin/misc_msg.cpp (hunk 1 above)
Comment (rsn8887) [reply SourceX/miniwin/misc_msg.cpp]: fixed in https://github.com/diasurgical/devilutionX/pull/181/commits/0ac63d03b0b3a444f5333dcad864bed7fa93eb03

File: Source/init.cpp [hunk 7]
```diff
@@ -164,6 +165,10 @@ void init_create_window(int nCmdShow)
 		nHeight = SCREEN_HEIGHT;
 	else
 		nHeight = GetSystemMetrics(SM_CYSCREEN);
+#else
+	nWidth = 640;
+	nHeight = 480;
```
Comment (AJenbo) [inline Source/init.cpp]: Might be better to either disable upscaling or modify GetSystemMetrics instead.

File: switch/readme-switch.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+- Install [devkitpro](https://devkitpro.org/wiki/Getting_Started#macOS)
+- Open Terminal
+- Type in ```dkp-pacman -S switch-freetype switch-mesa switch-glad switch-glm switch-sdl2 switch-sdl2_ttf switch-sdl2_mixer switch-libvorbis switch-libmikmod```
+- Type in ```make```
+
+- .nro lives in release. Test with an emulator (RyuJinx) or real hardware.
+
+### Credits
+
+- Initial Switch Port by MVG in 2019
```
Comment (mewmew) [inline switch/readme-switch.md]:
```suggestion
- Initial Switch Port by [MVG](https://github.com/lantus) in 2019
```

File: switch/mman.h (hunk 3 above)
Comment (rsn8887) [reply switch/mman.h]: fixed in https://github.com/diasurgical/devilutionX/pull/201/commits/2135690e21ec26830f9e2ee896c3eeb124366f8a

File: SourceX/miniwin/misc.cpp (hunk 4 above)
Comment (rsn8887) [reply SourceX/miniwin/misc.cpp]: fixed in https://github.com/diasurgical/devilutionX/pull/201/commits/13b16680628ae5da138176c9dbea6623641aa399

File: SourceX/miniwin/misc.cpp (hunk 5 above)
Comment (rsn8887) [reply SourceX/miniwin/misc.cpp]: Fixed in 
https://github.com/diasurgical/devilutionX/pull/201/commits/13b16680628ae5da138176c9dbea6623641aa399

in PR #201

File: switch/readme-switch.md (hunk 8 above)
Comment (rsn8887) [reply switch/readme-switch.md]: fixed in https://github.com/diasurgical/devilutionX/pull/201/commits/3c1ff1f3bc369347901c0b315e93d56f6b055517

in PR #201

File: Source/init.cpp (hunk 7 above)
Comment (rsn8887) [reply Source/init.cpp]: fixed in https://github.com/diasurgical/devilutionX/pull/201/commits/b988cf99db11a935a27790613db0f06d836e6598

in PR #201

**Comment by AJenbo** [comment]
I have taken care of most of the platform code. If anyone could test that it still loads correctly on there switch that would be a huge help:
https://1663-143324737-gh.circle-artifacts.com/0/diablo-nx.nro

**Comment by mewmew** [comment]
@AJenbo, do you need to do something specific to make it possible to run custom ROMs on the Switch?

**Comment by galaxyhaxz** [comment]
> to make it possible to run custom ROMs on the Switch?

As someone who originated from the console scene, I can say very much you aren't going to be running any custom code on a modern console without some sort of userland exploit. The Switch came with a design flaw that allowed them to be hacked to run unsigned code but Nintendo revised the motherboard so newer models cannot.

Modern consoles have sigchecks on code and certain keys are used to sign it based on desired user level access, with kernel keys being reserved for firmware devs and userland for game devs.

**Comment by AJenbo** [comment]
This port is now in fairly good shape, the only parts that are missing related to cleaning up the controller code.

It would also be great to get it compiling using CMake rather than a a custom make script @lantus is this something you have experience with?

**Comment by AJenbo** [comment]
Controller code has not been cleaned up and the build process moved to CMake, both thanks to @glebm.

All that remains is basically to see if `SetCursorPos` can be reenabled and a handful of other minor cleanups outlined in https://github.com/diasurgical/devilutionX/pull/443

**Comment by galaxyhaxz** [comment]
> All that remains is basically to see if SetCursorPos can be reenabled

Honestly I don't mind that one being disabled. Having the cursor move around when i show the inventory or spell list is quite annoying.

**Comment by AJenbo** [comment]
@galaxyhaxz well it is also needed for keeping the mouse in sync with the emulated mouses (touch and game controller), so would just be nice to have it work the same on all devices. @rsn8887 isn't it possible to attach mouse an keyboard to the Switch and play the game with that?

**Comment by rsn8887** [comment]
Yes that is possible, so... hmm... I guess the SDL mouse stuff should work on Switch. Maybe I was wrong.

**Comment by glebm** [comment]
@rsn8887 `SetCursorPos` calls `SDL_WrapMouseInWindow` which generates a mouse move event:

https://github.com/diasurgical/devilutionX/blob/2bb9aa4928dbe689d81f0ab1f81351e49c8c7ad9/SourceX/controls/plrctrls.cpp#L582

File: Packaging/switch/build.sh [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	build
+}
+
+build() {
+	mkdir -p ../../build
+	cd ../../build
+	rm -f CMakeCache.txt
+	DEVKITPRO="$DEVKITPRO" cmake .. -DBINARY_RELEASE=ON -DNONET=ON \
+		-DCMAKE_TOOLCHAIN_FILE=../CMake/switch/devkita64-libnx.cmake
+	DEVKITPRO="$DEVKITPRO" make -j "$(nproc)"
```
Comment (AJenbo) [inline Packaging/switch/build.sh]: Nproc is Linux only

File: .circleci/config.yml [hunk 10]
```diff
@@ -58,6 +58,20 @@ jobs:
       - run: cd build && cmake -DNIGHTLY_BUILD=ON -DCMAKE_TOOLCHAIN_FILE=../CMake/mingwcc.cmake ..
       - run: cd build && cmake --build . -j $(nproc)
       - store_artifacts: {path: ./build/devilutionx.exe, destination: devilutionx_x86.exe}
+  switch:
+    docker:
+      - image: devkitpro/devkita64:latest
```
Comment (AJenbo) [inline .circleci/config.yml]: Tool chain is outdated so won't work on firmware versions 9, either update at runtime or use a different image.

updating sdk using dkp-pacman should work

File: Packaging/switch/build.sh (hunk 9 above)
Comment (glebm) [reply Packaging/switch/build.sh]: How about `getconf _NPROCESSORS_ONLN`? It's Mac-compatible and hopefully works on msys2 windows as well.

File: SourceX/main.cpp [hunk 11]
```diff
@@ -25,5 +29,22 @@ static std::string build_cmdline(int argc, char **argv)
 int main(int argc, char **argv)
 {
 	auto cmdline = build_cmdline(argc, argv);
-	return dvl::WinMain(NULL, NULL, (char *)cmdline.c_str(), 0);
+
+#ifdef __SWITCH__
```
Comment (AJenbo) [inline SourceX/main.cpp]: Create init function in platform file and call atexit for the shutdown part.

**Pull request merged**